src/main/java/org/elasticsearch/common/util/MultiDataPathUpgrader.java

```diff
@@ -0,0 +1,366 @@
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.ImmutableSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.env.ShardLock;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.settings.IndexSettings;
+import org.elasticsearch.index.shard.*;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ */
+public class MultiDataPathUpgrader {
+
+    private final NodeEnvironment nodeEnvironment;
+    private final ESLogger logger = Loggers.getLogger(getClass());
+
+    public MultiDataPathUpgrader(NodeEnvironment nodeEnvironment) {
+        this.nodeEnvironment = nodeEnvironment;
+    }
+
+    /**
+     * Upgrades the given shard Id from multiple shard paths into the given target path.
+     *
+     * @see #pickShardPath(org.elasticsearch.index.shard.ShardId)
+     */
+    public void upgrade(ShardId shard, ShardPath targetPath) throws IOException {
+        final Path[] paths = nodeEnvironment.shardPaths(shard); // custom data path doesn't need upgrading
+        if (isTargetPathConfigured(paths, targetPath) == false) {
+            throw new IllegalArgumentException("shard path must be one of the shards data paths");
+        }
+        if (paths.length == 1) {
+            logger.info("{} only one data path configured - skipping upgrade", shard);
+            return;
+        }
+        logger.info("{} upgrading multi data dir to {}", shard, targetPath.getDataPath());
+        final ShardStateMetaData loaded = ShardStateMetaData.FORMAT.loadLatestState(logger, paths);
+        if (loaded == null) {
+            throw new IllegalStateException("Can't upgrade shard without shard state");
+        }
+        logger.info("{} load shard state {}", shard, loaded);
+
+        ShardStateMetaData.FORMAT.write(loaded, loaded.version, targetPath.getShardStatePath());
+        Files.createDirectories(targetPath.resolveIndex());
+        try (SimpleFSDirectory directory = new SimpleFSDirectory(targetPath.resolveIndex())) {
+            try (final Lock lock = directory.makeLock(IndexWriter.WRITE_LOCK_NAME)) {
+                if (lock.obtain(5000)) {
+                    upgradeFiles(shard, targetPath, targetPath.resolveIndex(), ShardPath.INDEX_FOLDER_NAME, paths);
+                } else {
+                    throw new IllegalStateException("Can't obtain lock on " + targetPath.resolveIndex());
+                }
+            }
+        }
+
+
+        upgradeFiles(shard, targetPath, targetPath.resolveTranslog(), ShardPath.TRANSLOG_FOLDER_NAME, paths);
+
+        logger.info("{} wipe old state files", shard);
```

Comment:
Why not wipe the entire source directories? I think it's good not to leave empty folders behind? we also leave lock files behind...